### ISSUE TITLE
client.next_result holds the GVL during a network call

### ISSUE THREAD

**Issue by jason-parkhurst** [brianmario/mysql2#1033]
When Mysql2::Client::MULTI_STATEMENTS is enabled and you call client.next_result, the GVL is held while we make a network call out to the MySQL server to obtain the next result set.  

rb_mysql_client_next_result calls mysql_next_result, which makes a call out to the MySQL server.

In the docs for mysql_next_result (https://dev.mysql.com/doc/refman/8.0/en/mysql-next-result.html) it's not totally clear that it reads from the network, though there is the hint that it can return CR_SERVER_GONE_ERROR or CR_SERVER_LOST.

In practice on Mac, it looks like it definitely does;  here's the stack which shows mysql_next_result making a network read call:
```
rb_mysql_client_next_result  (in mysql2.bundle) + 32  [0x10a4c18a0]  client.c:1142
 mysql_next_result  (in libmysqlclient.21.dylib) + 225  [0x10a4db959]
  cli_read_query_result(MYSQL*)  (in libmysqlclient.21.dylib) + 26  [0x10a4e50f8]
   cli_safe_read_with_ok  (in libmysqlclient.21.dylib) + 214  [0x10a4dd9ea]
    my_net_read(NET*)  (in libmysqlclient.21.dylib) + 94  [0x10a4e86d4]
     net_read_packet(NET*, unsigned long*)  (in libmysqlclient.21.dylib) + 47  [0x10a4e892b]
      net_read_raw_loop(NET*, unsigned long)  (in libmysqlclient.21.dylib) + 44  [0x10a4e8a61]
       vio_ssl_read(Vio*, unsigned char*, unsigned long)  (in libmysqlclient.21.dylib) + 47  [0x10a521082]
        ssl3_read_internal  (in libssl.1.0.0.dylib) + 96  [0x10aa157b8]
         ssl3_read_bytes  (in libssl.1.0.0.dylib) + 532  [0x10aa1849f]
          ssl3_read_n  (in libssl.1.0.0.dylib) + 587  [0x10aa1754b]
           BIO_read  (in libcrypto.1.0.0.dylib) + 100  [0x10aaf0fed]
            sock_read  (in libcrypto.1.0.0.dylib) + 50  [0x10aaf2a3f]
             read  (in libsystem_kernel.dylib) + 10  [0x7fff6566e142]
```             

In my case this caused an invisible deadlock (this call held the GVL and was waiting on a MySQL lock, and the Ruby thread with that lock couldn't run because it was in the same process and couldn't get the GVL).

**Comment by sodabrew** [comment]
This is the same bug #807 and #962 tried to fix back in 2016–2018, tracked down independently by two different reporters.

`client.next_result` calls `mysql_next_result()` synchronously, holding the GVL for however long the *next* result set in a multi-statement query takes to become ready. Two consequences:

- No other Ruby thread can run while it's waiting.
- It's not interruptible via `Thread#raise` — `Timeout.timeout` can't cancel it.

#807 (Ben Osheroff) wrapped the call in `rb_thread_call_without_gvl` plus a `rb_wait_for_single_fd` poll loop (the same pattern `do_query` already uses for the first query). The complication: libmysqlclient's `vio` layer can have already buffered the next result internally as a side effect of reading the first one — in which case polling the socket fd never sees a read-ready event, because the data's already sitting in the client library's memory, not on the wire. The workaround reached into `vio_priv.h` (not public API) to check `vio_has_data` before deciding whether to poll at all. #962 is a straight rebase of that same approach onto `master`.

Neither PR merged. From my own comment on #962 at the time: *"There are actually more possible states than my implementation accounts for. I'm not sure if there's a clean way to do this without the underlying methods exposed."* — i.e. I wasn't confident the vio-internals hack was correct/complete, not that the underlying bug wasn't real.

Checked current `master`: `rb_mysql_client_next_result` is still the original raw, blocking `mysql_next_result()` call, unchanged apart from an unrelated `affected_rows` cache added later. This is fully unfixed today.

This has come up again this year, too — #1457 (GVL-release-for-buffered-row-fetches) explicitly flags this issue as a related, out-of-scope gap: *"#1033 (`next_result` holds the GVL during a network call) is a different call site with a real network read; this PR only changes fetches from already-buffered results."*

Closing #962 as a historical artifact — the approach it took needs revisiting, not resurrecting as-is. Opening a new branch to look at this properly: a hang-reproducing test first, then a survey of what modern libmysqlclient/MariaDB Connector/C offer for this that we didn't have (or didn't know about) in 2016.
